### PR TITLE
Mgv6/cavegen: Add heightmap to keep large caves underground

### PR DESCRIPTION
**Pull request by paramat**
Hopefully reduces shadow bugs, see issue https://github.com/minetest/minetest/issues/1190

### CODE REVIEW

File: src/cavegen.cpp [hunk 1]
```diff
@@ -173,6 +173,36 @@ void CaveV6::makeTunnel(bool dirswitch) {
 		);
 	}
 
+	// Do not make large caves that are above ground.
+	// It is only necessary to check the startpoint and endpoint.
+	if (large_cave) {
+		v3s16 orpi(orp.X, orp.Y, orp.Z);
+		v3s16 veci(vec.X, vec.Y, vec.Z);
```
Comment (Zeno-) [inline src/cavegen.cpp]: What is a veci? :)

**Comment by paramat** [comment]
Not sure, code is copied from elsewhere in that file.

I just realised that in mgv6 the way large caves currently pass through the terrain surface should probably be left unchanged, as this shapes the terrain and mgv6 terrain must be preserved as it is. So closing for now.

**Comment by paramat** [comment]
The current code for keeping large caves underground aborts if either the start or end points of a cave are above ground. A possible compromise would be to abort only if both points are above ground, this might allow caves to pass through the surface while still removing caves that are completely above ground. Re-opening.

**Comment by paramat** [comment]
The code that keeps large caves below the heightmap is not effective when the cave is generated out into the mapblock shell (which happens often), so shadows would probably remain. Closing again.

**Comment by MT-Modder** [comment]
:frowning:

**Comment by paramat** [comment]
Fixed in a separate pull and merged https://github.com/minetest/minetest/commit/773aa8c57b70dd85273cb61286f8be577b4ca3aa

**Pull request closed without merging**